File: ruby/difference-of-squares/difference_of_squares.rb
class Squares
  private

  attr_reader :length

  public

  def initialize(length)
    @length = length
  end

  def square_of_sum
    (0..length).sum**2
  end

  def sum_of_squares
    (0..length).map { |n| n**2 }.sum
  end

  def difference
    square_of_sum - sum_of_squares
  end
end

File: typescript/difference-of-squares/difference-of-squares.ts
const add = (l: number, r: number) => l + r
const range = (length: number) => Array.from(Array(length + 1).keys())

export default class Squares {
  private readonly range: number[]
  constructor(length: number) {
    this.range = range(length)
  }

  get squareOfSum() {
    const sum = this.range.reduce(add, 0)
    return sum ** 2
  }

  get sumOfSquares() {
    const squares = this.range.map((n) => n ** 2)
    return squares.reduce(add, 0)
  }

  get difference() {
    return this.squareOfSum - this.sumOfSquares
  }
}
